File: internal/attachmentwire/transport_flow.go
package attachmentwire

import (
	"bytes"
	"fmt"
	"strconv"
)

// Flow acknowledgements pace server-to-browser output.
//
// Front-door WebSocket writes complete as soon as the bytes reach kernel and
// proxy buffers, so on a slow link everything written ahead of a liveness
// PONG — or of a WebSocket ping — waits in those buffers before it. The
// browser therefore acknowledges the attachment frames it has consumed, and
// the front door keeps only a bounded window of unacknowledged attachment
// bytes in flight. The count is cumulative over one WebSocket: "ACK 7" means
// the first seven attachment frames have been written into the terminal.
// Liveness and refusal frames are transport frames, not attachment frames,
// and are not counted. Browser to server only.
const TransportFlowPrefix = "PERSEA-FLOW/1 "

const transportFlowAck = "ACK "

// transportFlowDigits bounds the count to what a uint64 holds.
const transportFlowDigits = 20

// EncodeTransportFlowAck encodes one cumulative acknowledgement.
func EncodeTransportFlowAck(count uint64, direction Direction) ([]byte, error) {
	if direction != BrowserToServer || count == 0 {
		return nil, fmt.Errorf("transport flow protocol violation")
	}
	return []byte(TransportFlowPrefix + transportFlowAck + strconv.FormatUint(count, 10)), nil
}

// DecodeTransportFlowAck recognizes the reserved flow namespace without
// interpreting ordinary attachment frames. Once the prefix is present, the
// count must be a canonical positive decimal; any other syntax, or the wrong
// direction, is a protocol violation reported without payload material.
func DecodeTransportFlowAck(raw []byte, direction Direction) (uint64, bool, error) {
	if !bytes.HasPrefix(raw, []byte(TransportFlowPrefix)) {
		return 0, false, nil
	}
	violation := func() (uint64, bool, error) {
		return 0, true, fmt.Errorf("transport flow protocol violation")
	}
	body := raw[len(TransportFlowPrefix):]
	if !bytes.HasPrefix(body, []byte(transportFlowAck)) {
		return violation()
	}
	digits := body[len(transportFlowAck):]
	if direction != BrowserToServer || len(digits) == 0 || len(digits) > transportFlowDigits || digits[0] == '0' {
		return violation()
	}
	for _, digit := range digits {
		if digit < '0' || digit > '9' {
			return violation()
		}
	}
	count, err := strconv.ParseUint(string(digits), 10, 64)
	if err != nil {
		return violation()
	}
	return count, true, nil
}

File: ui/src/transport_flow.ts
// Flow acknowledgements pace server-to-browser output. The front door keeps a
// bounded window of unacknowledged attachment frames in flight, so liveness
// and control frames never queue behind more than that window on a slow link.
// The page acknowledges a frame once everything it caused in the terminal has
// been written, never on arrival: arrival only proves the bytes left the
// network, not that the page kept up. The count is cumulative per WebSocket
// and covers attachment frames only; liveness and refusal frames are not
// counted.
export const TRANSPORT_FLOW_PREFIX = "PERSEA-FLOW/1 ";

export function encodeBrowserFlowAck(count: number): string {
  if (!Number.isSafeInteger(count) || count < 1) throw new Error("transport flow count unavailable");
  return `${TRANSPORT_FLOW_PREFIX}ACK ${count}`;
}

// FlowAcknowledger numbers the attachment frames of one WebSocket and
// acknowledges the longest prefix of them that has been consumed. Frames can
// complete out of order (a control frame completes at once, output only after
// the terminal wrote it), so only a contiguous prefix is acknowledged.
// Acknowledgements are coalesced: consumption within one task sends one.
export class FlowAcknowledger {
  private received = 0;
  private consumed = 0;
  private acknowledged = 0;
  private readonly completed = new Set<number>();
  private scheduled = false;
  private retired = false;

  constructor(
    // Sends one acknowledgement; false when the socket can no longer carry it.
    private readonly send: (payload: string) => boolean,
    private readonly schedule: (callback: () => void) => void = (callback) => queueMicrotask(callback),
  ) {}

  // Numbers the next attachment frame.
  receive(): number {
    return ++this.received;
  }

  // Marks one numbered frame consumed.
  consume(sequence: number): void {
    if (this.retired || sequence <= this.consumed || sequence > this.received || this.completed.has(sequence)) return;
    this.completed.add(sequence);
    while (this.completed.delete(this.consumed + 1)) this.consumed++;
    if (this.consumed > this.acknowledged && !this.scheduled) {
      this.scheduled = true;
      this.schedule(() => this.flush());
    }
  }

  retire(): void {
    this.retired = true;
    this.completed.clear();
  }

  private flush(): void {
    this.scheduled = false;
    if (this.retired || this.consumed <= this.acknowledged) return;
    if (this.send(encodeBrowserFlowAck(this.consumed))) this.acknowledged = this.consumed;
  }
}
